Firestore/core/test/firebase/firestore/model/field_path_test.cc

```diff
@@ -0,0 +1,272 @@
+  const FieldPath a{"rooms"};
+  const FieldPath empty;
+  const FieldPath abc_dupl{"rooms", "Eros", "messages"};
+
+  EXPECT_EQ(ab, abc.DropLast());
+  EXPECT_EQ(a, abc.DropLast().DropLast());
+  EXPECT_EQ(empty, abc.DropLast().DropLast().DropLast());
+}
+
+TEST(FieldPath, Concatenation) {
+  const FieldPath path;
+  const FieldPath a{"rooms"};
+  const FieldPath ab{"rooms", "Eros"};
+  const FieldPath abc{"rooms", "Eros", "messages"};
+
+  EXPECT_EQ(a, path.Concat("rooms"));
+  EXPECT_EQ(ab, path.Concat("rooms").Concat("Eros"));
+  EXPECT_EQ(abc, path.Concat("rooms").Concat("Eros").Concat("messages"));
+  EXPECT_EQ(abc, path.Concat(FieldPath{"rooms", "Eros", "messages"}));
+  EXPECT_EQ(abc, path.Concat({"rooms", "Eros", "messages"}));
+
+  const FieldPath bcd{"Eros", "messages", "this_week"};
+  EXPECT_EQ(bcd, abc.DropFirst().Concat("this_week"));
+}
+
+TEST(FieldPath, Comparison) {
+  const FieldPath abc{"a", "b", "c"};
+  const FieldPath abc2{"a", "b", "c"};
+  const FieldPath xyz{"x", "y", "z"};
+  EXPECT_EQ(abc, abc2);
+  EXPECT_NE(abc, xyz);
+
+  const FieldPath empty;
+  const FieldPath a{"a"};
+  const FieldPath b{"b"};
+  const FieldPath ab{"a", "b"};
+
+  EXPECT_TRUE(empty < a);
+  EXPECT_TRUE(a < b);
+  EXPECT_TRUE(a < ab);
+
+  EXPECT_TRUE(a > empty);
+  EXPECT_TRUE(b > a);
+  EXPECT_TRUE(ab > a);
+}
+
+TEST(FieldPath, IsPrefixOf) {
+  const FieldPath empty;
+  const FieldPath a{"a"};
+  const FieldPath ab{"a", "b"};
+  const FieldPath abc{"a", "b", "c"};
+  const FieldPath b{"b"};
+  const FieldPath ba{"b", "a"};
+
+  EXPECT_TRUE(empty.IsPrefixOf(empty));
+  EXPECT_TRUE(empty.IsPrefixOf(a));
+  EXPECT_TRUE(empty.IsPrefixOf(ab));
+  EXPECT_TRUE(empty.IsPrefixOf(abc));
+  EXPECT_TRUE(empty.IsPrefixOf(b));
+  EXPECT_TRUE(empty.IsPrefixOf(ba));
+
+  EXPECT_FALSE(a.IsPrefixOf(empty));
+  EXPECT_TRUE(a.IsPrefixOf(a));
+  EXPECT_TRUE(a.IsPrefixOf(ab));
+  EXPECT_TRUE(a.IsPrefixOf(abc));
+  EXPECT_FALSE(a.IsPrefixOf(b));
+  EXPECT_FALSE(a.IsPrefixOf(ba));
+
+  EXPECT_FALSE(ab.IsPrefixOf(empty));
+  EXPECT_FALSE(ab.IsPrefixOf(a));
+  EXPECT_TRUE(ab.IsPrefixOf(ab));
+  EXPECT_TRUE(ab.IsPrefixOf(abc));
+  EXPECT_FALSE(ab.IsPrefixOf(b));
+  EXPECT_FALSE(ab.IsPrefixOf(ba));
+
+  EXPECT_FALSE(abc.IsPrefixOf(empty));
+  EXPECT_FALSE(abc.IsPrefixOf(a));
+  EXPECT_FALSE(abc.IsPrefixOf(ab));
+  EXPECT_TRUE(abc.IsPrefixOf(abc));
+  EXPECT_FALSE(abc.IsPrefixOf(b));
+  EXPECT_FALSE(abc.IsPrefixOf(ba));
+}
+
+TEST(FieldPath, AccessFailures) {
+  const FieldPath path;
+  ASSERT_DEATH_IF_SUPPORTED(path.front(), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.back(), "");
+  ASSERT_DEATH_IF_SUPPORTED(path[0], "");
+  ASSERT_DEATH_IF_SUPPORTED(path[1], "");
+  ASSERT_DEATH_IF_SUPPORTED(path.at(0), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropFirst(), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropFirst(2), "");
+  ASSERT_DEATH_IF_SUPPORTED(path.DropLast(), "");
+}
+
+TEST(FieldPath, Parsing) {
+  const auto parse = [](const std::pair<std::string, size_t> expected) {
```

Comment:
Objective C uses macros for this. Helper functions are a little more verbose, but (I hope, haven't tried) mapping failure to line number should be easier.